### PR TITLE
tests: Preserve data dir from failed linearizability tests

### PR DESCRIPTION
**Pull request by serathius**
cc @ahrtr

### CODE REVIEW

File: tests/linearizability/linearizability_test.go [hunk 1]
```diff
@@ -192,3 +185,39 @@ type trafficConfig struct {
...
+			err := os.RemoveAll(memberDataDir)
+			if err != nil {
+				t.Error(err)
+			}
+			t.Logf("saving %s data dir to %q", member.Config().Name, memberDataDir)
+			err = os.Rename(member.Config().DataDirPath, memberDataDir)
+			if err != nil {
+				t.Error(err)
+			}
+		}
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:212]: add` t.Error("Model is not linearizable")`

File: tests/linearizability/linearizability_test.go [hunk 2]
```diff
@@ -192,3 +185,39 @@ type trafficConfig struct {
...
+	}
+	t.TempDir()
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:214]: ?
```suggestion

```

File: tests/linearizability/linearizability_test.go [hunk 3]
```diff
@@ -104,22 +105,14 @@ func testLinearizability(ctx context.Context, t *testing.T, config e2e.EtcdProce
...
 	if err != nil {
 		t.Error(err)
 	}
-	err = porcupine.VisualizePath(etcdModel, info, path)
-	if err != nil {
-		t.Errorf("Failed to visualize, err: %v", err)
+	linearizable := checkOperationsAndPersistResults(t, operations, clus)
+	if linearizable != porcupine.Ok {
+		t.Error("Model is not linearizable")
 	}
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:115]:
```suggestion
	checkOperationsAndPersistResults(t, operations, clus)
```

File: tests/linearizability/linearizability_test.go [hunk 4]
```diff
@@ -192,3 +185,39 @@ type trafficConfig struct {
...
 }
+
+func checkOperationsAndPersistResults(t *testing.T, operations []porcupine.Operation, clus *e2e.EtcdProcessCluster) porcupine.CheckResult {
+	path, err := filepath.Abs(filepath.Join(resultsDirectory, strings.Replace(t.Name(), "/", "_", -1)))
+	if err != nil {
+		t.Error(err)
+	}
+	err = os.MkdirAll(path, 0700)
+	if err != nil {
+		t.Error(err)
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:196]: `t.Error(err)` means that the test will continue to execute .

```suggestion
		t.Fatal(err)
```

File: tests/linearizability/linearizability_test.go [hunk 5]
```diff
@@ -192,3 +185,39 @@ type trafficConfig struct {
 	clientCount int
 	traffic     Traffic
 }
+
+func checkOperationsAndPersistResults(t *testing.T, operations []porcupine.Operation, clus *e2e.EtcdProcessCluster) porcupine.CheckResult {
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:189]:
```suggestion
func checkOperationsAndPersistResults(t *testing.T, operations []porcupine.Operation, clus *e2e.EtcdProcessCluster)  {
```

File: tests/linearizability/linearizability_test.go [hunk 6]
```diff
@@ -192,3 +185,39 @@ type trafficConfig struct {
...
+	}
+	t.TempDir()
+
+	visualizationPath := filepath.Join(path, "history.html")
+	t.Logf("saving visualization to %q", visualizationPath)
+	err = porcupine.VisualizePath(etcdModel, info, visualizationPath)
+	if err != nil {
+		t.Errorf("Failed to visualize, err: %v", err)
+	}
+	return linearizable
```
Comment (ahrtr) [inline tests/linearizability/linearizability_test.go:222]:
```suggestion

```

**Comment by ahrtr** [comment]
Please consider to use `github.com/stretchr/testify` to simplify the checking on test results.

File: tests/linearizability/linearizability_test.go (hunk 4 above)
Comment (serathius) [reply tests/linearizability/linearizability_test.go:196]: Will still want to report model not linearizable even if writing report will not work.

**Comment by serathius** [comment]
> Please consider to use github.com/stretchr/testify to simplify the checking on test results.

Sure, maybe as a followup PR. I would want to first understand how looks testify addoption throughout other tests.

File: tests/linearizability/linearizability_test.go (hunk 2 above)
Comment (ahrtr) [reply tests/linearizability/linearizability_test.go:214]: You resolved this comment without providing any explanation. Why do you add this?

**Review by ahrtr — APPROVED** [review]
LGTM

**Pull request merged**